Added combined nt

dandeduck opened the pull request.

dandeduck committed:
rearranged  ntprop factory

dandeduck committed:
removed lines after class start

dandeduck committed:
implemented combined nt thing

dandeduck commented:
So, about tests... Junit 4 or 5? also conventions? anything else I need to know?

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/beans/NtDoubleProperty.java:
```diff
@@ -8,7 +8,6 @@
 import java.util.Objects;
 
 public class NtDoubleProperty extends NtPropertyBase implements DoubleProperty {
-
```
What are you trying to be declined?
Put those lines back.

dandeduck replied:
;-;

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/combined/CombinedNtProperties.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class CombinedNtProperties {
```
Weird name... call it something like `NtTable`

dandeduck replied:
k

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/combined/CombinedNtProperties.java:
```diff
@@ -0,0 +1,66 @@
+        table = tableInstance.getTable(name);
+    }
+
+    public void addBooleanEntry(String name, boolean initialValue) {
```
It's more `put` than `add`, if we're using `map` terminology

dandeduck replied:
Not sure if it should be in the map terminology now...

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/combined/CombinedNtProperties.java:
```diff
@@ -0,0 +1,66 @@
+        table = tableInstance.getTable(name);
+    }
+
+    public void addBooleanEntry(String name, boolean initialValue) {
```
Didn't you want some sugar syntax? Something builder like?

dandeduck replied:
It's planned, yes

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/combined/CombinedNtProperties.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public NetworkTableEntry getEntry(String name) {
+        return NtEntryMap.get(name);
```
Just query the table.... No need to save it again in your own map.

dandeduck replied:
yeahyeahyeah

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/combined/CombinedNtProperties.java:
```diff
@@ -0,0 +1,66 @@
+
+public class CombinedNtProperties {
+    private final Map<String, NetworkTableEntry> NtEntryMap;
+    private final NetworkTable table;
```
`mTable`

dandeduck replied:
;-;

tomtzook reviewed (changes requested):
Imma smack you

dandeduck committed:
brought back the lines

dandeduck committed:
deleted combined thing shit

dandeduck committed:
returned NtTable and added NtTableBuilder

dandeduck commented:
- I changed how NtTable works and added the builder. Do you think that the builder should get the table in the ctor or as a method like other things? In other words, should the builder be specific to every `build()` ?

- Also, I don't really like using the table object in NtTable (besides the confusing naming) it's not really immutable and just adds a stupid wrapper to it all... I have some ideas to fix this, but none of them are really good or pretty. That's why I wanted to use a stand-alone map from the beginning.

dandeduck commented:
> Imma smack you

Yes please, daddy

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/combined/NtTable.java:
```diff
@@ -0,0 +1,47 @@
+package com.flash3388.frc.nt.combined;
+
+import edu.wpi.first.networktables.*;
```
Don't use global imports, they're problematic

dandeduck replied:
it was the IDE I SWEAR

tomtzook replied:
You can remove it in _settings_

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/combined/NtTable.java:
```diff
@@ -0,0 +1,47 @@
+         this(tableInstance.getTable(name));
+    }
+
+    public NtTable(NetworkTable table) {
```
Ctor convention in Flashlib stuff is opposite: general contructor first, overloads bellow.

dandeduck replied:
BUT WHY?

tomtzook replied:
Because I said so

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/combined/NtTableBuilder.java:
```diff
@@ -0,0 +1,50 @@
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.networktables.NetworkTableValue;
+
+public class NtTableBuilder {
```
Flashlib convention for builders is to make them as inner classes for the class they build.

dandeduck replied:
I kinda like that, but only from the user perspective

tomtzook reviewed (changes requested):
>     * I changed how NtTable works and added the builder. Do you think that the builder should get the table in the ctor or as a method like other things? In other words, should the builder be specific to every `build()` ?
> 
>     * Also, I don't really like using the table object in NtTable (besides the confusing naming) it's not really immutable and just adds a stupid wrapper to it all... I have some ideas to fix this, but none of them are really good or pretty. That's why I wanted to use a stand-alone map from the beginning.

- Table in builder ctor. 
- What do you mean by "_specific for every `build`_"?
- I don't really understand you objection to using `NetworkTable`? You're using the same table, how is a `map` better here?

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/combined/NtTable.java:
```diff
@@ -0,0 +1,47 @@
+package com.flash3388.frc.nt.combined;
```
Can just be in `nt` package. I don't really understand the `combined` naming. It's also barely a thing.

dandeduck replied:
Yeah, my bad

dandeduck committed:
changes to fit conventions

dandeduck committed:
one more

dandeduck commented:
> > ```
> > * I changed how NtTable works and added the builder. Do you think that the builder should get the table in the ctor or as a method like other things? In other words, should the builder be specific to every `build()` ?
> > 
> > * Also, I don't really like using the table object in NtTable (besides the confusing naming) it's not really immutable and just adds a stupid wrapper to it all... I have some ideas to fix this, but none of them are really good or pretty. That's why I wanted to use a stand-alone map from the beginning.
> > ```
> 
>     * Table in builder ctor.
> 
>     * What do you mean by "_specific for every `build`_"?
> 
>     * I don't really understand you objection to using `NetworkTable`? You're using the same table, how is a `map` better here?

- By that I mean that every instance of the builder can build just one "type" of the object, because if you can't change the table, then the created objects
-My points was that I want to create an object that contains constant entries, but when I am using a table, you can always add to it, thus making it kinda useless. The point is for it to be fit for some purpose, e.g. PID tunning.

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/NtTable.java:
```diff
@@ -0,0 +1,91 @@
+        return getEntry(name).getValue();
+    }
+
+    public class Builder {
```
Should have mentioned that it should be at the beginning of the class.

dandeduck replied:
Yeah, should have. Sometimes I feel like you don't tell me stuff deliberately so that you will delay the PR longer

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/NtTable.java:
```diff
@@ -0,0 +1,91 @@
+            this(NetworkTableInstance.getDefault(), name);
+        }
+
+        public Builder(NetworkTableInstance tableInstance, String name) {
```
This should be second

dandeduck replied:
What's the convention exactly?

tomtzook replied:
Class cascade upwards

dandeduck replied:
Is it like "used method below the user method" but wrong and reverse?

dandeduck replied:
lol wrote without refreshing, but still got it

dandeduck replied:
How can something cascade upwards?

tomtzook replied:
Gravity

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/NtTable.java:
```diff
@@ -0,0 +1,91 @@
+        this(NetworkTableInstance.getDefault(), name);
+    }
+
+    public NtTable(NetworkTableInstance tableInstance, String name) {
```
should be second

dandeduck replied:
may be

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/NtTable.java:
```diff
@@ -0,0 +1,91 @@
+         this(tableInstance.getTable(name));
+    }
+
+    public boolean getBooleanEntryValue(String name) {
```
You don't actually need to say "entry". Shorter names can be nice.

dandeduck replied:
Wanted to be safe, but whateves

tomtzook commented:
> By that I mean that every instance of the builder can build just one "type" of the object, because if you can't change the table, then the created objects

No idea what you mean by that. What "type"? _datatype_? Why would you limit it by data types?

> My points was that I want to create an object that contains constant entries, but when I am using a table, you can always add to it, thus making it kinda useless. The point is for it to be fit for some purpose, e.g. PID tunning.

I can understand noticing that it can be changed for somewhere else. But I don't think it's a problem. It's just a utility, not some secure storage.

dandeduck committed:
fixed for conventions

dandeduck committed:
changed naming a bit

dandeduck commented:
Why didn't you resolve the things that I fixed from previous cr?

tomtzook commented on flashlib.frc.nt/src/main/java/com/flash3388/frc/nt/NtTable.java:
```diff
@@ -0,0 +1,91 @@
+            return addEntry(name, NetworkTableValue.makeBoolean(initialValue));
+        }
+
+        public Builder(NetworkTableInstance tableInstance, String name) {
```
I'm questioning whether or not it was on purpose

dandeduck replied:
may be

tomtzook replied:
Dude, above the `Builder(String)` one:

```Java
        public Builder(NetworkTable table) {
            mTable = table;
        }

        public Builder(NetworkTableInstance tableInstance, String name) {
            this(instance.getTable(name));
        }

        public Builder(String name) {
            this(NetworkTableInstance.getDefault(), name);
        }
```

dandeduck replied:
oh really?

dandeduck committed:
coscading upwards is like shitting upwards

dandeduck committed:
hmmm

tomtzook committed:
Merge branch 'development' into improving-nt
